### PR TITLE
Fix for "poetry lock --no-update" so that dependencies with extras are not updated unnecessarily

### PR DESCRIPTION
**Pull request by radoering**
# Pull Request Check List

Resolves: #4612
Resolves: #3128
Resolves: #5185

- [x] Added **tests** for changed code.
- [ ] Updated **documentation** for changed code.

When calling `poetry lock` with `--no-update` dependencies with extras should not be updated unnecessarily.

### CODE REVIEW

**Comment by radoering** [comment]
@neersighted: Shall I rebase the branch so that the FreeBSD test can succeed or should I wait for your review?

**Comment by neersighted** [comment]
> 
> 
> @neersighted: Shall I rebase the branch so that the FreeBSD test can succeed or should I wait for your review?

Please go ahead and rebase! There are enough changes in master that significantly stale PRs will need to be rebased to have confidence in testing anyway. Thanks!

File: src/poetry/mixology/version_solver.py
Comment (neersighted) [inline src/poetry/mixology/version_solver.py:391]: I feel like this line is a bit clunky -- do you think this is cleaner than a `None` check?

File: src/poetry/mixology/version_solver.py
Comment (neersighted) [inline src/poetry/mixology/version_solver.py:384]: There's not a good way to unroll the nested logic here, is there? This is a bit dense, but I'm not seeing an obvious simplification myself.

**Review by neersighted — CHANGES_REQUESTED** [review]

File: src/poetry/mixology/version_solver.py
Comment (radoering) [reply src/poetry/mixology/version_solver.py:391]: Probably, a None check is easier to understand. I will change it.

File: src/poetry/mixology/version_solver.py
Comment (radoering) [reply src/poetry/mixology/version_solver.py:384]: If you want to refrain from taking advantage of the fact that the list is sorted and thus from the early return if the version is not found, you can simplify the loop as follows:

```
if locked is not None:
    for version in packages:
        if version.version == locked.version:
            break
```

Comment (neersighted) [reply src/poetry/mixology/version_solver.py:384]: In that case I'm wondering if `next()` wouldn't be clearer -- e.g.

```python
    version = next((p.version for p in packages if p.version == locked.version), None)
```

Thoughts?

Comment (radoering) [reply src/poetry/mixology/version_solver.py]: I like it, but it has to be:

`version = next((p for p in packages if p.version == locked.version), None)`

Maybe a bit confusing that the variable `version` is a `package`. Should we rename it to `package`?

Comment (radoering) [reply src/poetry/mixology/version_solver.py]: @neersighted Sorry for bothering, but I had the impression that the review of this PR has already been well advanced - maybe, the naming of the variable being the only point to be clarified before merging. (Correct me if I'm wrong.) I'd appreciate if you find some minutes so that we may finalize it.

**Review by neersighted — APPROVED** [review]

**Pull request merged**
